### PR TITLE
experimentation: make APIs stricter and better documented

### PR DESCRIPTION
**Pull request by Augustyniak**
### Description

1. Look at all of the changes in `experimentation.proto`. Basically, codifying some of the requirements that were already in place but were not stated anywhere in public docs.
2. Add `reason` parameter to `CancelExperimentRunRequest` and start enforcing it.
3. Modify experimentation frontend so that it provides `reason` parameter when a user cancels an experiment run.
4. Start displaying `termination_reason` in experiment run details view.

### Testing Performed
1. Manual creation and cancellation of server experiments.

### CODE REVIEW

File: backend/service/chaos/experimentation/experimentstore/experiment_run_details_test.go [hunk 1]
```diff
@@ -12,108 +12,79 @@ import (
...
-}
-
-func TestCanceledExperiment(t *testing.T) {
-	logger := zaptest.NewLogger(t).Sugar()
-
-	end := sql.NullTime{Valid: false}
-	cancellation := sql.NullTime{
-		Time:  startTime.AddDate(0, 0, -1),
-		Valid: true,
+func TestExperimentRunDetailsStatus(t *testing.T) {
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/experimentstore/experiment_run_details_test.go:15]: sam tests, just cleaned up

File: backend/service/chaos/experimentation/experimentstore/experiment_specification.go [hunk 2]
```diff
@@ -30,8 +27,6 @@ func NewExperimentSpecification(ced *experimentationv1.CreateExperimentData, now
 	// If the run identifier is not provided, generate one
 	if runId == "" {
 		runId = id.NewID().String()
-	} else if !regexp.MustCompile(runIdRegExp).MatchString(runId) {
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/experimentstore/experiment_specification.go:33]: this is enforced by proto validation rules now

File: backend/service/chaos/experimentation/experimentstore/storer.go [hunk 3]
```diff
@@ -228,10 +228,11 @@ func (s *storer) GetListView(ctx context.Context) ([]*experimentation.ListViewIt
 		var details string
 		run := ExperimentRun{}
 		config := ExperimentConfig{Config: &any.Any{}}
-		err = rows.Scan(&run.Id, &run.StartTime, &run.EndTime, &run.CancellationTime, &run.creationTime, &config.id, &details)
+		err = rows.Scan(&run.Id, &run.StartTime, &run.EndTime, &run.CancellationTime, &run.CreationTime, &config.id, &details)
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/experimentstore/storer.go:231]: now sure how it continues to work with `*time.Time` as opposed to `*sql.nulltime`

File: backend/service/chaos/experimentation/experimentstore/experiment_run.go [hunk 4]
```diff
@@ -70,22 +70,35 @@ func (er *ExperimentRun) CreateProperties(now time.Time) ([]*experimentation.Pro
...
+			properties = append(properties, &experimentation.Property{
+				Id:    "canceled_at",
+				Label: "Canceled At",
+				Value: cancellationTimeTimestamp,
+			})
+		}
+
+		terminationReason := "Unknown"
+		if er.TerminationReason != "" {
+			terminationReason = er.TerminationReason
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/experimentstore/experiment_run.go:95]: start displaying termination reason if test was cancelled or stopped

File: api/chaos/experimentation/v1/create_experiment_data.proto [hunk 5]
```diff
@@ -15,7 +15,7 @@ message CreateExperimentData {
   // Its max length is limited to 100 characters. The identifier is supposed to be user-readable and
   // URL renderable - for this reason, allowed characters are limited to English characters, digits and
   // the following special characters: "-", ".", "_" and "~".
-  string run_id = 1;
+  string run_id = 1 [ (validate.rules).string.pattern = "^[A-Za-z0-9-._~]*$" ];
```
Comment (snowp) [inline api/chaos/experimentation/v1/create_experiment_data.proto:18]: max length as well?

File: api/chaos/experimentation/v1/experimentation.proto [hunk 6]
```diff
@@ -35,19 +36,24 @@ message CreateOrGetExperimentResponse {
...
     STATUS_UNSPECIFIED = 0;
+    // Return experiments that are currently running.
     STATUS_RUNNING = 1;
   }
 
-  string config_type = 1;
+  // Return experiments whose config's type URL matches the provided config type URL.
+  string config_type = 1 [ (validate.rules).string = {min_bytes : 1} ];
+  // Return experiments whose status is equal to the provided value.
   Status status = 2;
```
Comment (snowp) [inline api/chaos/experimentation/v1/experimentation.proto:57]: Should this be required?

File: api/chaos/experimentation/v1/experimentation.proto [hunk 7]
```diff
@@ -61,7 +67,10 @@ message CancelExperimentRunRequest {
     pattern : "{id}"
   };
 
-  string id = 1;
+  // The identifier of an experiment run to cancel.
+  string id = 1 [ (validate.rules).string = {min_bytes : 1} ];
+  // The reason for the cancelation request.
+  string reason = 2 [ (validate.rules).string = {min_bytes : 1, max_len : 32} ];
```
Comment (snowp) [inline api/chaos/experimentation/v1/experimentation.proto:73]: Thoughts on bumping the reason limit?

File: api/chaos/experimentation/v1/experimentation.proto [hunk 8]
```diff
@@ -109,27 +119,32 @@ service ExperimentsAPI {
...
   }
+  // Get the list of experiments in the form that's optimized for displaying to the end user.
   rpc GetListView(GetListViewRequest) returns (GetListViewResponse) {
     option (google.api.http) = {
       post : "/v1/chaos/experimentation/getListView"
       body : "*"
     };
     option (clutch.api.v1.action).type = READ;
   }
+  // Fetch the list of properties in the format that's optimized for displaying to the end user.
```
Comment (snowp) [inline api/chaos/experimentation/v1/experimentation.proto:147]: Just a note but it's a bit odd to have endpoints that are tailored for rendering and for general lookups without any naming differences. Seems to be conflating a regular CRUD API with generating UI elements. 

Maybe these should be separated in some way, e.g. name prefix/suffix, different API, etc.?

Out of scope of this PR, just a thought

File: api/chaos/experimentation/v1/experimentation.proto (hunk 7 above)
Comment (Augustyniak) [reply api/chaos/experimentation/v1/experimentation.proto:73]: I can bump it 👍  - I did not want to introduce a database migration in my PR but it blew up anyway so I may as well introduce it here.

File: api/chaos/experimentation/v1/experimentation.proto (hunk 8 above)
Comment (Augustyniak) [reply api/chaos/experimentation/v1/experimentation.proto:147]: Ah, that makes sense - I will follow up on this

File: api/chaos/experimentation/v1/experimentation.proto (hunk 7 above)
Comment (Augustyniak) [reply api/chaos/experimentation/v1/experimentation.proto]: increased to 100

File: api/chaos/experimentation/v1/experimentation.proto (hunk 6 above)
Comment (Augustyniak) [reply api/chaos/experimentation/v1/experimentation.proto:57]: I could not find required rule for enums. Am I missing something

File: api/chaos/experimentation/v1/create_experiment_data.proto (hunk 5 above)
Comment (Augustyniak) [reply api/chaos/experimentation/v1/create_experiment_data.proto]: added

File: api/chaos/experimentation/v1/experimentation.proto (hunk 6 above)
Comment (snowp) [reply api/chaos/experimentation/v1/experimentation.proto]: you do like "defined only" and "not = unspecified" or something like that

Comment (Augustyniak) [reply api/chaos/experimentation/v1/experimentation.proto]: yea, I've already added `Status status = 2 [ (validate.rules).enum.defined_only = true ];`

File: api/chaos/experimentation/v1/experimentation.proto [hunk 9]
```diff
@@ -35,20 +36,25 @@ message CreateOrGetExperimentResponse {
...
+    // Return experiments that are currently running.
     STATUS_RUNNING = 1;
   }
 
-  string config_type = 1;
-  Status status = 2;
+  // Return experiments whose config's type URL matches the provided config type URL.
+  string config_type = 1 [ (validate.rules).string = {min_bytes : 1} ];
+  // Return experiments whose status is equal to the provided value.
+  Status status = 2 [ (validate.rules).enum.defined_only = true ];
```
Comment (snowp) [inline api/chaos/experimentation/v1/experimentation.proto:57]: Defined only also allows setting UNSPECIFIED, do we want to allow that?

Comment (Augustyniak) [reply api/chaos/experimentation/v1/experimentation.proto]: added `not_int: 0` requirement

**Review by snowp — APPROVED** [review]

**Pull request merged**